pkg/querier/worker/scheduler_processor.go

```diff
@@ -213,16 +215,30 @@ func (sp *schedulerProcessor) runRequest(ctx context.Context, logger log.Logger,
 			Body: []byte(errMsg),
 		}
 	}
+	var c client.PoolClient
+	var retries int
 
-	c, err := sp.frontendPool.GetClientFor(frontendAddress)
-	if err == nil {
+	for {
+		c, err = sp.frontendPool.GetClientFor(frontendAddress)
+		if err != nil {
+			break
+		}
 		// Response is empty and uninteresting.
 		_, err = c.(frontendv2pb.FrontendForQuerierClient).QueryResult(ctx, &frontendv2pb.QueryResultRequest{
 			QueryID:      queryID,
 			HttpResponse: response,
 			Stats:        stats,
 		})
+		if err == nil || retries >= maxNotifyFrontendRetries {
+			break
+		}
+		// If the used connection returned and error, remove it from the pool and retry.
+		level.Warn(logger).Log("msg", "retrying to notify frontend about finished query", "err", err, "frontend", frontendAddress, "retries", retries)
+
+		sp.frontendPool.RemoveClientFor(frontendAddress)
```

Comment:
@ortuman do you remember the reason for removing the connection here? If there are multiple concurrent queries which get an error, remove the client, then try to get a new client, they will remove each other's new clients.

Reply:
I think removing the client isn't strictly necessary. Opened a PR to remove them https://github.com/grafana/mimir/pull/6713, but happy to hear feedback and change it

Reply:
IIRC, idea was to get rid of client that may have its connection closed, but wasn't yet removed from the pool.

>  If there are multiple concurrent queries which get an error, remove the client, then try to get a new client, they will remove each other's new clients.

Yes, we haven't considered this. Your proposal to remove specific client only (if it still matches) would improve this behaviour.